src/riak_kv_compile_tab.erl

```diff
@@ -22,37 +22,66 @@
 
 -module(riak_kv_compile_tab).
 
--export([is_compiling/1]).
--export([get_state/1]).
--export([get_ddl/1]).
--export([new/1]).
--export([insert/4]).
--export([update_state/2]).
+-export([
+         delete_dets/1,
+         get_compiled_ddl_version/1,
+         get_ddl/1,
+         get_state/1,
+         get_tables_needing_recompiling/1,
+         insert/5,
+         is_compiling/1,
+         new/1,
+         update_state/2]).
 
 -define(TABLE, ?MODULE).
 
--type compiling_state() :: compiling | compiled | failed.
+-type compiling_state() :: compiling | compiled | failed | retrying.
 -export_type([compiling_state/0]).
 
 -define(is_compiling_state(S),
         (S == compiling orelse
          S == compiled orelse
-         S == failed)).
+         S == failed orelse
+         S == retrying)).
 
 %% 
 -spec new(file:name()) ->
         {ok, dets:tab_name()} | {error, any()}.
 new(FileDir) ->
     FilePath = filename:join(FileDir, [?TABLE, ".dets"]),
-    dets:open_file(?TABLE, [{type, set}, {repair, force}, {file, FilePath}]).
+    lager:debug("Opening DDL DETS table ~s", [FilePath]),
+    Result = dets:open_file(?TABLE, [{type, set}, {repair, force}, {file, FilePath}]),
+    %% Only update the DETS database if it opens successfully
+    case Result of
+        {ok, _} ->
+            %% Convert all legacy DDL compilation records to current format
+            upgrade_legacy_records(),
+
+            %% Clean up any lingering tables stuck in the compiling state
```

Comment:
Two nitpicks: suggest using "records" instead of "tables" since the latter is a bit ambiguous in a DETS context.

Also would like to see an explanation of why these compiling records are lingering (node reboot during compilation). Same with the function comment.